package.json

```diff
@@ -7,7 +7,7 @@
     "test": "test"
   },
   "scripts": {
-    "test": "npm run lint && mocha --recursive test",
+    "test": "npm run lint && mocha --exit",
```

Comment:
Okay, I will remove the `--exit` flag and update the mocha tests as you've described.

Good catch on the hanging bug. Why is the `Server` instance not being destroyed? Shouldn't it be deleted once resolve is called in `manager.instance(...)?`? What would you like to do about the hanging server instance?

Here's a way to upgrade to v3.0 of MongoDB in Travis CI: [https://github.com/travis-ci/travis-ci/issues/3694#issuecomment-155186560](https://github.com/travis-ci/travis-ci/issues/3694#issuecomment-155186560)

Reply:
I experience hanging after the tests are completed in MongoDB 4.0.1.

Reply:
This is a problem with the `instance` method in general, it returns a `Server` (think of this as a direct connection mongodb client) which the caller needs to explicitly destroy themselves. I think it's a bad API, and I think we could do better by providing the method I described above, which would implicitly destroy the instance after the passed in lambda is complete.

Reply:
We have tooling to install multiple versions of mongodb, we just need to update the travis scripts to use them. I think we should punt on that until we resolve the rest of your issues. I made [NODE-1803](https://jira.mongodb.org/browse/NODE-1803) to track the issue.

Reply:
Also, it's difficult to understand your previous comment: are you still experiencing the hang with MongoDB 4.0.1?

Reply:
@mbroadst I'm not sure if it works with 4.0.1 because there are some other issues fixed by my PRs that are getting in the way. I think it is working though.

Whoops, I accidently added support for MongoDB 3.2. Is this the general idea of how you want to update MongoDB on Travis CI?

Reply:
Looks like it's present with 3.2 still:

[https://travis-ci.org/mongodb-js/mongodb-topology-manager/builds/469575450](https://travis-ci.org/mongodb-js/mongodb-topology-manager/builds/469575450)

Reply:
Yeah there's another bug here it looks like this line: https://github.com/mongodb-js/mongodb-topology-manager/blob/master/lib/replset.js#L555. This should be checking `result.version` for the version comparison. This is a case of very poorly tested and named code paths unfortunately.

Reply:
👍 The test works now with my method of installing MongoDB. I saw what @imlucas said and so I will remove my code to pave the way for `mongodb-version-manager` to be used.

The last piece is to deal with the `instance` function. Your opinion is to rename the function to `withInstance` or `usingInstance` and accept a function as an argument that will be called. The `configuration` function would change to look like the first excerpt of code.

I wanted to discuss if it would be better to just have the user call `.destroy()` on the instance before finishing with it. The second excerpt of code shows how the `configuration` function would change.

I prefer the second option because it is a minor change to make and it sticks more to the `co` style of Promises. The first option is using a callback so that is one more indentation level for code to be at. As another point, since the function is no longer in a co-wrapped generator function, there would be one more indentation level to get it back to where yielding Promises is allowed.

Just wanted to plead my case first. I'll let you make the decision and I can implement either option quickly.

```javascript
          manager.instance(credentials, function(server) {
            // Get the configuration document
            var cursor = server.cursor('local.system.replset', {
              find: 'local.system.replset',
              query: {},
              limit: 1
            });

            // Execute next
            cursor.next(function(err, d) {
              if (err) return reject(err);
              if (!d) return reject(new Error('no replicaset configuration found'));
              resolve(d);
            });
          });
```

```javascript
          // Get a server instance
          var server = yield manager.instance(credentials);
          // Get the configuration document
          var cursor = server.cursor('local.system.replset', {
            find: 'local.system.replset',
            query: {},
            limit: 1
          });

          // Execute next
          cursor.next(function(err, d) {
            server.destroy(); // CHANGES MADE HERE
            if (err) return reject(err);
            if (!d) return reject(new Error('no replicaset configuration found'));
            resolve(d);
          });
```

Reply:
Let's leave discussion of a potential `withInstance`/`usingInstance` to a later PR, I think there's enough going on here for now 😄